### DISCUSSION TITLE
Log database insert/delete/update and what user made the change

### DISCUSSION THREAD

**Discussion by enesien** [supabase/supabase#17758] [Questions] [answered]
How can I effectively log database changes (insert/update/delete) and identify the authenticated user responsible for the change? 

I've explored options like pgaudit and triggers. Triggers seem to suit my needs best as they allow me to log specific actions based on the table and selectively for certain tables. However, my primary requirement is to accurately identify **who** initiated the change.

For instance, if employee Bob modifies a contract status from unpaid to paid, I aim to create a log entry such as:

`Bob changed contract 1234 from paid to unpaid`

or

`Employee 4321 changed contract 1234 from paid to unpaid`

Is it possible to achieve this using triggers or other methods, and how can I implement it effectively?

Cheers.

**Comment by GaryAustin1** [comment, marked as answer]
You have access to auth.uid() in triggers and that is the best place to do this.  You can generate a table that you insert to as your log in the trigger.

**Reply by enesien** [reply]
Ahh great. So do I have to pass the uid from the trigger to the function as an argument or do functions have access to auth.uid(), as well?

**Reply by GaryAustin1** [reply]
They do.  You can get to any jwt claim in any function.  Auth.uid() and Auth.jwt() are the main ways.

**Reply by enesien** [reply]
Great, thanks!

**Reply by GaryAustin1** [reply]
You do have to run the function from an original trigger function or an rpc call to the function chain, otherwise claims are null.  Obviously if a user did not initiate thru the database API there are no claims set.

**Comment by arjunlol** [comment]
You can use something like https://github.com/BemiHQ/bemi for this. It plugs into PG and your application to automatically format lower-level logs with application context (e.g. user, apiEndpoint, etc) and store in a way that's easily queryable.
